src/common/os_auto_flush_windows.c

```diff
@@ -35,17 +35,195 @@
  */
 
 #include <windows.h>
+#include <inttypes.h>
 #include "out.h"
 #include "os.h"
+#include "endian.h"
+#include "os_auto_flush_windows.h"
 
 /*
- * os_auto_flush -- check if platform supports auto flush
- * XXX - for now Windows implementation always returns 0
+ * is_nfit_available -- (internal) check if platform supports NFIT table.
  */
 int
-os_auto_flush(void)
+is_nfit_available()
 {
-	LOG(15, NULL);
+	LOG(3, "is_nfit_available()");
+
+	DWORD signatures_size;
+	char *signatures = NULL;
+	int is_nfit = 0;
+	DWORD offset = 0;
+
+	signatures_size = EnumSystemFirmwareTables(ACPI_SIGNATURE, NULL, 0);
+	if (signatures_size == 0) {
+		ERR("!EnumSystemFirmwareTables");
+		goto err;
+	}
+	signatures = (char *)Malloc(signatures_size + 1);
+	if (signatures == NULL) {
+		ERR("!malloc");
+		goto err;
+	}
+	int ret = EnumSystemFirmwareTables(ACPI_SIGNATURE,
+					signatures, signatures_size);
+	signatures[signatures_size] = '\0';
+	if (ret != signatures_size || ret == 0) {
+		ERR("!EnumSystemFirmwareTables");
+		goto err;
+	}
+
+	while (offset <= signatures_size) {
+		int nfit_sig  = strncmp(signatures + offset,
+				NFIT_STR_SIGNATURE, NFIT_SIGNATURE_LEN);
+		if (nfit_sig == 0) {
+			is_nfit = 1;
+			goto end;
+		}
+		offset += NFIT_SIGNATURE_LEN;
+	}
+
+end:
+	Free(signatures);
+	return is_nfit;
+
+err:
+	if (signatures)
+		Free(signatures);
+	return -1;
+}
+
+/*
+ * check_capabilities -- (internal) check if specific capabilities bits are set.
+ *
+ * ACPI 6.2A Specification:
+ * Bit[0] - CPU Cache Flush to NVDIMM Durability on
+ * Power Loss Capable. If set to 1, indicates that platform
+ * ensures the entire CPU store data path is flushed to
+ * persistent memory on system power loss.
+ * Bit[1] - Memory Controller Flush to NVDIMM Durability on Power Loss Capable.
+ * If set to 1, indicates that platform provides mechanisms to automatically
+ * flush outstanding write data from the memory controller to persistent memory
+ * in the event of platform power loss. Note: If bit 0 is set to 1 then this bit
+ * shall be set to 1 as well.
+ */
+static int
+check_capabilities(uint32_t capabilities)
+{
+	LOG(3, "check_capabilities capabilities 0x%" PRIx32, capabilities);
+
+	int CPU_cache_flush = CHECK_BIT(capabilities, 0);
+	int memory_controller_flush = CHECK_BIT(capabilities, 1);
+
+	LOG(15, "CPU_cache_flush %d, memory_controller_flush %d",
+			CPU_cache_flush, memory_controller_flush);
+	if (memory_controller_flush == 1 && CPU_cache_flush == 1)
+		return 1;
 
 	return 0;
 }
+
+/*
+ * parse_nfit_buffer -- (internal) parse nfit buffer
+ * if platform_capabilities struct is available fill up pc structure.
+ */
+static void
+parse_nfit_buffer(unsigned char *nfit_buffer, unsigned long buffer_size,
```

Comment:
this function should fail if it didn't write anything to the platform_capabilities buffer because otherwise the caller won't be able to tell he can actually read from it.